aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,231 @@
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+
+namespace at {
+namespace native {
+
+// helper
+namespace {
+
+template <typename scalar_t>
+std::vector<int64_t> sort_indices(const TensorAccessor<scalar_t, 1>& v) {
+  std::vector<int64_t> indices(v.size(0));
+  std::iota(indices.begin(), indices.end(), 0);
+  std::stable_sort(
+      indices.begin(),
+      indices.end(),
+      [&](const scalar_t& i1, const scalar_t& i2) { return v[i1] < v[i2]; });
+  return indices;
+}
+
+template <size_t N>
+std::vector<int64_t> indices2csr(
+    const TensorAccessor<int64_t, N>& indices,
+    int64_t dim) {
+  auto nnz = indices.size(0);
+  std::vector<int64_t> csr(dim + 1, int64_t(0));
+  int64_t last_i = 0;
+  for (int64_t index = 0; index < nnz; index++) {
+    int64_t i = indices[index];
+    if (i == last_i) {
+      csr[last_i + 1] += 1;
+    } else {
+      for (int64_t j = last_i; j < i + 1; j++) {
+        csr[j + 1] = csr[last_i + 1];
+      }
+      last_i = i;
+      csr[last_i + 1] += 1;
+    }
+  }
+  for (int64_t j = last_i; j < dim; j++) {
+    csr[j + 1] = csr[last_i + 1];
+  }
+  return csr;
+}
+
+template <typename scalar_t>
+void sparse_matmul_kernel(
+    Tensor& result,
+    const Tensor& self,
+    const Tensor& other) {
+  auto indices_a = self._indices().contiguous();
+  auto values_a = self._values().contiguous();
+
+  auto indices_b = other._indices().contiguous();
+  auto values_b = other._values().contiguous();
+
+  auto nnz_a = values_a.size(0);
+  auto nnz_b = values_b.size(0);
+  auto indices_accessor_a = indices_a.accessor<int64_t, 2>();
+  auto indices_accessor_b = indices_b.accessor<int64_t, 2>();
+
+  auto values_accessor_a = values_a.accessor<scalar_t, 1>();
+  auto values_accessor_b = values_b.accessor<scalar_t, 1>();
+
+  auto result_indices = result._indices().contiguous();
+  auto result_values = result._values().contiguous();
+
+  std::map<std::pair<int64_t, int64_t>, scalar_t> d;
+  if (self.is_coalesced() && other.is_coalesced()) {
+    for (int64_t n1 = 0; n1 < nnz_a; n1++) {
+      auto r2 = indices2csr(indices_accessor_b[0], other.size(0));
+      for (int n2 = r2[indices_accessor_a[1][n1]];
+           n2 < r2[indices_accessor_a[1][n1] + 1];
+           n2++) {
+        auto current_index = std::make_pair(
+            indices_accessor_a[0][n1], indices_accessor_b[1][n2]);
+        d[current_index] += values_accessor_a[n1] * values_accessor_b[n2];
+      }
+    }
+  } else {
+    for (int64_t n1 = 0; n1 < nnz_a; n1++) {
+      for (int n2 = 0; n2 < nnz_b; n2++) {
+        if (indices_accessor_b[0][n2] == indices_accessor_a[1][n1]) {
+          auto current_index = std::make_pair(
+              indices_accessor_a[0][n1], indices_accessor_b[1][n2]);
+          d[current_index] += values_accessor_a[n1] * values_accessor_b[n2];
+        }
+      }
+    }
+  }
+  int64_t nnz_result = d.size();
+  std::vector<int64_t> values_size = {nnz_result};
+  result_indices.resize_({2, nnz_result});
```

Comment:
This should take care of the continuity of the indices and values tensors (as long as their original size doesn't match these sizes).